pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java

```diff
@@ -421,20 +421,43 @@ public static class QueryOptionKey {
         public static final String USE_SCAN_REORDER_OPTIMIZATION = "useScanReorderOpt";
         public static final String MAX_EXECUTION_THREADS = "maxExecutionThreads";
 
-        /** Number of groups AggregateOperator should limit result to after sorting.
-         *  Trimming happens only when (sub)query contains order by and limit clause. */
+        /**
+         * Number of groups AggregateOperator should limit result to after sorting.
+         * Trimming happens only when (sub)query contains order by and limit clause.
+         */
         public static final String GROUP_TRIM_SIZE = "groupTrimSize";
 
-        /** Number of groups GroupByOperator should limit result to after sorting.
-         * Trimming happens only when (sub)query contains order by clause. */
+        /**
+         * Number of groups GroupByOperator should limit result to after sorting.
+         * Trimming happens only when (sub)query contains order by clause.
+         */
         public static final String MIN_SEGMENT_GROUP_TRIM_SIZE = "minSegmentGroupTrimSize";
 
-        /** Max number of groups GroupByCombineOperator (running at server) should return .*/
+        /**
+         * Max number of groups GroupByCombineOperator (running at server) should return .
+         */
         public static final String MIN_SERVER_GROUP_TRIM_SIZE = "minServerGroupTrimSize";
 
-        /** Max number of groups GroupByDataTableReducer (running at broker) should return. */
+        /**
+         * Max number of groups GroupByDataTableReducer (running at broker) should return.
+         */
         public static final String MIN_BROKER_GROUP_TRIM_SIZE = "minBrokerGroupTrimSize";
 
+        /**
+         * Number of threads used in the server level final reduce, this is used for expensive aggregation functions,
+         * and we want to push down reduce to server level instead of streaming all the data back to broker for
+         * global reduce.
+         *
+         * E.g. Funnel queries are considered as expensive aggregation functions.
+         */
+        public static final String NUM_THREADS_FOR_SERVER_FINAL_REDUCE = "numThreadsForServerFinalReduce";
```

Comment:
Consider `numThreadsExtractFinalResult`. Based on my code reading, it will also be applied to broker reduce. Same for chunk size

Reply:
done